ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -642,40 +642,48 @@ forkBlockForging IS{..} blockForging =
 
         -- Add the block to the chain DB
         let noPunish = InvalidBlockPunishment.noPunishment   -- no way to punish yourself
-        result <- lift $ ChainDB.addBlockAsync chainDB noPunish newBlock
-        -- Block until we have processed the block
-        curTip <- lift $ atomically $ ChainDB.blockProcessed result
-
-        -- Check whether we adopted our block
-        when (curTip /= blockPoint newBlock) $ do
-          isInvalid <- lift $ atomically $
-            ($ blockHash newBlock) . forgetFingerprint <$>
-            ChainDB.getIsInvalidBlock chainDB
-          case isInvalid of
-            Nothing ->
-              trace $ TraceDidntAdoptBlock currentSlot newBlock
-            Just reason -> do
-              trace $ TraceForgedInvalidBlock currentSlot newBlock reason
-              -- We just produced a block that is invalid according to the
-              -- ledger in the ChainDB, while the mempool said it is valid.
-              -- There is an inconsistency between the two!
-              --
-              -- Remove all the transactions in that block, otherwise we'll
-              -- run the risk of forging the same invalid block again. This
-              -- means that we'll throw away some good transactions in the
-              -- process.
-              lift $ removeTxs mempool (map (txId . txForgetValidated) txs)
-          exitEarly
-
-        -- We successfully produced /and/ adopted a block
+        -- Make sure that if an async exception is thrown while a block is
+        -- added to the chain db, we will remove txs from the mempool.
         --
-        -- NOTE: we are tracing the transactions we retrieved from the Mempool,
-        -- not the transactions actually /in the block/. They should always
-        -- match, if they don't, that would be a bug. Unfortunately, we can't
-        -- assert this here because the ability to extract transactions from a
-        -- block, i.e., the @HasTxs@ class, is not implementable by all blocks,
-        -- e.g., @DualBlock@.
-        trace $ TraceAdoptedBlock currentSlot newBlock txs
+        -- 'addBlockAsync' is a non-blocking action, so `mask_` would suffice,
+        -- but the finalizer is a blocking operation, hence we need to use
+        -- 'uninterruptibleMask_' to make sure that async exceptions do not
+        -- interrupt it.
+        uninterruptibleMask_ $ do
+          result <- lift $ ChainDB.addBlockAsync chainDB noPunish newBlock
+          -- Block until we have processed the block
+          curTip <- lift $ atomically $ ChainDB.blockProcessed result
```

Comment:
I think this block is assuming that `blockProcessed` is guaranteed to become unblocked.

But I'm not sure that's true. In particular, I doubt the corresponding ChainDB code is blocking async exceptions. So, suppose:

- We call `addBlockAsync`.
- The ChainDB shuts down (eg the users requests shutdown, eg a fatal error, etc) before unblocking `blockProcessed`.

That would leave this thread deadlocked, wouldn't it?

... maybe that's OK? This thread doesn't own any resources (eg that would need some logic to close gracefully). So this thread would only die when the main GHC thread ends (ie when the OS reaps the process).

I _think_ the ChainDB only has a single lifetime (ie it only shuts down if the node is shutting down), so `blockProcessed` will only deadlock if the node is shutting down---ie there's no risk we'll accumulate a bunch of zombie threads.

Reply:
However, opting into this stuck thread behavior seems like asking for trouble in the future.

The alternative is to not mask async exns here. It seems acceptable to allow for the thread to die between adding the block and being notified that it was processed. The continuation only matters for i) completeness of log message and ii) removing the offending transactions from the mempool in the unlikely case that a bug has caused the node to forge a block that it then identifies as invalid. Unless the operator is rotating their block credentials with high frequency, those same txs would probably be included in the next block, and so would get caught by the continuation then.

So, the options and their downsides are:

- include the mask: we are opting-in to some possible unusual stuck-thread behavior when the ChainDB shutsdown while/immediately after the node forges a block

- exclude the mask: when the user signals a credentials change immediately after the node forges an invalid block (which would require additional bug somewhere in the forge logic), the node would miss two consecutive forging opportunities instead of just the first.

- ??? is there a Pareto-dominant option I didn't consider here?

cc @coot

Reply:
Ah, nevermind. https://hackage.haskell.org/package/base-4.17.0.0/docs/Control-Exception.html#t:BlockedIndefinitelyOnSTM If we have the mask (option 1, as on your current diff), then I think the RTS will raise `BlockedIndefinitelyOnSTM` (*synch*ronously, _I think_?) in this situation if the ChainDB shutsdown while the thread is waiting on `blockProcessed`. It involves more moving parts than I'd like, but there all being used how they're intended.

Reply:
> synchronously, I think?

Does it matter how the exception is raised? I'd expect that in this case such case RTS sets the thread to throw an exception, and does not wait for it to be raised (like the `throwTo` does).

Anyway relaying on `BlockedIndefinitelyOnSTM` makes me shiver :wink:. 

But all this discussion points to a problem with the current approach.  The thread which actually adds block should inform the other end of the queue that it died, e.g. `blockProcessed` ought to return `Maybe (Point blk)`, if `Nothing` is returned the block producing thread would know that the thread that was adding a block died and it could terminate as well, without being blocked.

Reply:
> Does it matter how the exception is raised? I'd expect that in this case such case RTS sets the thread to throw an exception, and does not wait for it to be raised (like the `throwTo` does).

Well, if the spec is that it would be raised _synchronously_ (eg it's considered as if the STM transaction itself decided to raise it), then I'd be confident that the masking state wouldn't the thread from receiving it while stuck in the STM transaction. That's all I was thinking.

-----

> But all this discussion points to a problem with the current approach.

Hmm, yes, I like that solution (Pareto-dominant! 👏).

Reply:
Great. Yes, looks good. I'm Resolving this Thread, since I just started some new Threads for the new code.